File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
# 유기농 배추
from collections import deque
import sys
input = sys.stdin.readline


t = int(input())

def bfs(graph, x, y):
    q = deque()
    q.append((x,y))
    graph[y][x] = 0
    dx = [-1,1,0,0]
    dy = [0,0,-1,1]
    while q:
        x,y = q.popleft()
        # graph[y][x] = 0   #방문처리
        # cabbage.remove((x,y))
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx<0 or nx>=m or ny<0 or ny>=n:
                continue
            if graph[ny][nx] == 0:
            # if (x,y) not in cabbage:
                continue
            else:
                q.append((nx,ny))
                cabbage.remove((nx,ny))
                graph[ny][nx] = 0

for _ in range(t):

    m, n, k = map(int,input().split())
    graph = [[0 for _ in range(m)] for _ in range(n)]
    cabbage = []
    for _ in range(k):
        x, y = map(int,input().split())
        cabbage.append((x,y))
        graph[y][x] = 1

    
    for x, y in cabbage:
        bfs(graph, x, y)
    print(len(cabbage))

File: 백준/Silver/1012. 유기농 배추/유기농 배추.js
const readline = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];

readline
  .on("line", function (line) {
    input.push(line);
  })
  .on("close", function () {
    solution(input);
    process.exit();
  });

function solution(input) {
  const T = Number(input[0]);
  const inputs = input.slice(1).map((row) => row.split(" ").map(Number));
  for (let cycle = 0; cycle < T; cycle++) {
    const [M, N, K] = inputs.shift();
    const graph = Array.from({ length: N }, () => new Array(M).fill(0));
    for (let i = 0; i < K; i++) {
      const [x, y] = inputs.shift();
      graph[y][x] = 1;
    }

    let cnt = 0;
    for (let y = 0; y < N; y++) {
      for (let x = 0; x < M; x++) {
        if (graph[y][x] === 1) {
          dfs([x, y], graph, N, M);
          cnt++;
        }
      }
    }
    console.log(cnt);
  }

  function dfs(start, graph, N, M) {
    const [cx, cy] = start;
    graph[cy][cx] = 0;
    const dx = [0, 0, -1, 1];
    const dy = [-1, 1, 0, 0];
    for (let i = 0; i < dx.length; i++) {
      const [nx, ny] = [cx + dx[i], cy + dy[i]];
      if (nx < 0 || ny < 0 || nx >= M || ny >= N || graph[ny][nx] === 0) {
        continue;
      }
      graph[ny][nx] = 0;
      dfs([nx, ny], graph, N, M);
    }
  }
}
